File: python/socketd/socketd/transport/stream/StreamMangerDefault.py
import logging

from socketd.transport.stream.Stream import StreamInternal
from socketd.transport.stream.StreamManger import StreamManger
from socketd.utils.MapUtils import MapUtils

logger = logging.getLogger(__name__)


class StreamMangerDefault(StreamManger):
    def __init__(self, config):
        self.config = config
        self.stream_map:[str, StreamInternal] = {}

    def get_stream(self, sid):
        return self.stream_map.get(sid)

    def add_stream(self, sid, stream):
        if stream.demands() == 0:
            # Zero demands, do not add
            return

        self.stream_map[sid] = stream

        # Add stream timeout handling as a backup insurance
        stream_timeout = stream.timeout() if stream.timeout() > 0 else self.config.get_stream_timeout()
        if stream_timeout > 0:
            stream.insurance_start(self, stream_timeout)

    def remove_stream(self, sid):
        stream: StreamInternal = MapUtils.remove(self.stream_map, sid)
        if stream:
            stream.insurance_cancel()
            logger.debug(f"{self.config.get_role_name()} stream removed, sid={sid}")

File: javascript/socketd/src/socketd/transport/stream/impl/StreamMangerDefault.ts
import {Config} from "../../core/Config";
import {Asserts} from "../../core/Asserts";
import {Constants} from "../../core/Constants";
import {StreamInternal} from "../Stream";
import {StreamManger} from "../StreamManger";

export class StreamMangerDefault implements StreamManger{
    _config:Config;
    _streamMap: Map<string, StreamInternal<any>>

    constructor(config:Config) {
        this._config = config;
        this._streamMap = new Map<string, StreamInternal<any>>();
    }

    /**
     * 获取流接收器
     *
     * @param sid 流Id
     */
    getStream(sid:string) {
        const tmp = this._streamMap.get(sid);
        if (tmp) {
            return tmp;
        } else {
            return null;
        }
    }

    /**
     * 添加流接收器
     *
     * @param sid    流Id
     * @param stream 流
     */
    addStream(sid, stream: StreamInternal<any>) {
        Asserts.assertNull("stream", stream);

        if(stream.demands() == Constants.DEMANDS_ZERO){
            //零需求，则不添加
            return;
        }

        this._streamMap.set(sid, stream);

        //增加流超时处理（做为后备保险）
        const streamTimeout = stream.timeout() > 0 ? stream.timeout() : this._config.getStreamTimeout();
        if (streamTimeout > 0) {
            stream.insuranceStart(this, streamTimeout);
        }
    }

    /**
     * 移除流接收器
     *
     * @param sid 流Id
     */
    removeStream(sid) {
        const stream = this.getStream(sid);

        if (stream) {
            this._streamMap.delete(sid);
            stream.insuranceCancel();
            console.debug(`${this._config.getRoleName()} stream removed, sid=${sid}`);
        }
    }
}
